### ISSUE TITLE
naive cycle variant

### ISSUE THREAD

**Issue by dvanhorn** [plum-umd/monadic-eval#14]
I'd like the following variant: detects cycles and produces the empty set.  This should have a cache of seen states and produce empty set if asked to eval a state in the cache.

(It's pretty urgent that this and #13 get done ASAP.  Can someone claim this and get it done soon?)

**Comment by dvanhorn** [comment]
PS: and it can't be @davdar.

**Comment by labichn** [comment]
En route.

**Comment by philnguyen** [comment]
Thanks @labichn I would probably have been clueless for this one

**Comment by labichn** [comment]
This is a bit underspecified, what is the expected result for the following expressions?

```
(eval (parse '(+ 5 (+ 5 err)))) => (list (failure)
                                         ∅
                                         {set (list '(+ 5 (+ 5 err)) ∅ ∅)
                                              (list 5 ∅ ∅)
                                              (list '(+ 5 err) ∅ ∅)})
                      OR?       => (list {set} 
                                         ∅
                                         {set (list '(+ 5 (+ 5 err)) ∅ ∅)
                                              (list 5 ∅ ∅)
                                              (list '(+ 5 err) ∅ ∅)})
                      OR?       => {set}
```

The last is really non-standard for our setup. The second has us in a NondetT and requires a top-level `eval` to throw away all results we've gotten so far. The first is the most natural to implement (just adding a `hijack` inside `ev-reach`.

**Comment by dvanhorn** [comment]
I think the first since there's no loop in this program.

**Comment by labichn** [comment]
But you'll hit the configuration `(list 5 ∅ ∅)` twice, won't you? Or is there a way to recognize an infinite cycle that I'm not seeing, unless we're caching?

**Comment by dvanhorn** [comment]
I guess what I'm suggesting is to have an accumulator for seen states that you add to every time you recur.  If the current state is in the accumulator, you return empty set.  Something like the following:

```
(define (eval e r s seen)
  (if (in? seen (list e r s))
      (set)
      (match e
        [(num n) (set n)]
        [(app e0 e1)
         (let ((v0s (eval e0 r s (cons (list e r s) seen)))
               (v1s (eval e1 r s (cons (list e r s) seen))))
            ...)]))))
```

So while it's true you see 5 twice you don't see 5 while trying to eval 5.

I want something that in the concrete semantics is a sound cycle detector.  What you're proposing is not.

**Comment by labichn** [comment]
How do you write the predicate `trying-to-eval-5?` without a defunctionalized stack? I don't see how the code block you posted does this. We would need to cache the stack in addition to the seen set to do this, wouldn't we?

**Comment by philnguyen** [comment]
Can't we just use a reader to accumulate configurations we've seen for the particular path?

**Comment by davdar** [comment]
if you make the cache path-sensitive rather than flow-insensitive you might get what you're looking for.

**Comment by labichn** [comment]
I think the implicit goal of no caching (apart from seen states) applies like in #13, otherwise we're just giving the `eval-pdcfa^` `alloc-bang^`.

**Comment by labichn** [comment]
That said, `eval-pdcfa^`, `alloc-bang^`, and the concrete delta would do what you're looking for, I think. But we'd still be caching.

**Comment by dvanhorn** [comment]
@labichn : I didn't have to write any predicates in my sketch of the solution.  I agree it should something as simple as @philnguyen says.

**Comment by davdar** [comment]
in my mind it should be analogous to having a cache ($ : exp \* env \* store -> val), except you track a seen cache (seen$ : exp \* env \* store -> bool), and you make this cache path-sensitive, so resulting evaluator is (exp \* env \* store -> set(val \* store \* seen$)) rather than (exp \* env \* store -> (set (val \* store) \* $)

**Comment by labichn** [comment]
The above code would halt while evaluating the second `(num 5)` in `(+ 5 (+ 5 err))`, wouldn't it?. The states would be identical. They wouldn't include any other subexpressions, just `(list (num 5) ∅ ∅)`. Or is something more subtle happening?

**Comment by dvanhorn** [comment]
No it would not.  Notice that you never add things like (5 ...) to the seen set because there's no recursive call when evaluating 5.

**Comment by davdar** [comment]
I see, I think Phil was on the right track with the reader idea. To mondaify DVH's example you need to use local when adding something to the seen set, or else the stuff seen while evaluating e0 will leak into the seen set when evaluating e1. Or you can keep it a state and use an analogous "localize" operator for the state monad.

**Comment by dvanhorn** [comment]
@davdar : I think of caches as carrying the result you want when something is in the table, which is not what I'm suggesting.  But otherwise sounds right.

**Comment by labichn** [comment]
Had my mind stuck on Writer--got it now. Sorry for the confusion!

**Comment by davdar** [comment]
yeah, cache was the wrong word

**Comment by dvanhorn** [comment]
Closed by 799d4ee4
